core/utils/lazy.go

```diff
@@ -0,0 +1,43 @@
+package utils
+
+import (
+	"sync"
+)
+
+type LazyLoad[T any] struct {
+	f     func() (T, error)
+	state T
+	lock  sync.RWMutex
+	once  sync.Once
+}
+
+func NewLazyLoad[T any](f func() (T, error)) *LazyLoad[T] {
+	return &LazyLoad[T]{
+		f: f,
+	}
+}
+
+func (l *LazyLoad[T]) Get() (out T, err error) {
+
+	// fetch only once (or whenever cleared)
+	l.lock.Lock()
+	l.once.Do(func() {
+		l.state, err = l.f()
+	})
+	l.lock.Unlock()
+
+	// if err, clear so next get will retry
+	if err != nil {
+		l.Reset()
+	}
+
+	l.lock.RLock()
+	defer l.lock.RUnlock()
+	return l.state, err
```

Comment:
I think you can drop this lock by assigning `out = l.state` up above while the lock is already held.
```suggestion
	return
```

Reply:
ah good point...using a read lock at this point doesn't matter since there's a write lock right before it so that would block until all reads are done anyway